fix(challenge): handle console and network errors (#528)

Silvitui opened the pull request:
This **PR** completes task **528**  part of user story 516:  "Fix console errors and verify if there are any related issues in the Network tab"

## ✅ Console Errors Fixed : 

### 1. `NG0100: ExpressionChangedAfterItHasBeenCheckedError`
This error occurs when Angular detects a value change **after** the view has already been checked.  
It happened because we were trying to update the view with data that changed too late in the lifecycle.

#### ✅ How we fixed it:

We used `ChangeDetectorRef.detectChanges()` to notify Angular **manually** that a value has changed and the view needs to be updated accordingly.
---
### 2. `Cannot read properties of undefined (reading 'timesFavorited')`
This happened because we tried to access a property (`timesFavorited`) **before the `challenge` data had finished loading** from the backend.

#### ✅ How we fixed it:
We added a conditional check in the template to make sure the `challenge` object exists before displaying its content.

### 3. `innerHTML` Security Warning
There was a warning due to the use of `innerHTML`, which can be risky because of potential XSS attacks.  
We replaced it with **Angular interpolation** (`{{ }}`), which is safer and prevents injection of malicious HTML.

---

## ⚠️ Backend Error (409 Conflict)

When clicking the **"Send solution"** button, a `409 Conflict` error appeared in the console:
We are aware of this issue and currently working on it.

## 🐞 Missing Route Error in Console

When navigating to **“Technical Tests”** under the “Challenges” section, the following error appears:
This happens because the `componentToBuild` route/component **does not exist yet**.  
The error will disappear once the corresponding component is created and registered in the Angular router.

---

Silvitui committed:
fix(challenge): handle console errors

Victorreca commented on src/app/modules/challenge/components/challenge/challenge.component.html:
Please adjust the spacing to match the formatting used in the develop branch for consistency.

Silvitui replied:
Okey! It's already done

Victorreca reviewed (commented):
Everything looks good overall! Just a small adjustment needed in challenge.component.html to align the spacing with the develop branch. Great job!

Silvitui committed:
style(challenge-header): fix spaces for consistency

Victorreca reviewed (approved):
Looks good to me, approved ✅

ivilarop commented:
@Silvitui 

🔧 1. CHANGELOG and version
This PR fixes functional and visible errors that directly affect the component's behavior, so it should:

Include an entry in CHANGELOG.md

Include a version change according to the wiki's versioning guidelines (in this case, a PATCH increment would suffice)

📌 2. Lack of definition of the change *ngIf="!solutionSent"
The description details the motivation for the fixed errors (innerHTML, detectChanges(), undefined data), but doesn't explain the logic change:

*ngIf="solutionSent" → *ngIf="!solutionSent"

This change doesn't seem to respond directly to a console or network error like the others, but rather to a functional or usability improvement. It would have been advisable to separate it into another PR or at least justify it in the description.

💡 For similar future cases, separating functional fixes from technical bugs can help improve traceability, review, and deployment control.

ivilarop commented on src/app/modules/challenge/components/challenge-info/challenge-info.component.html:
```diff
@@ -128,7 +129,7 @@ <h3>
 
       <!-- Si la solución no ha sido enviada, mostrar las opciones de cancelación y aceptación -->
-      <div class="w-100 button d-md-none" *ngIf="solutionSent">
+      <div class="w-100 button d-md-none" *ngIf="!solutionSent">
```
The description details the motivation for the fixed errors (innerHTML, detectChanges(), undefined data), but doesn't explain the logic change:

*ngIf="solutionSent" → *ngIf="!solutionSent"

This change doesn't seem to respond directly to a console or network error like the others, but rather to a functional or usability improvement. It would have been advisable to separate it into another PR or at least justify it in the description.

💡 For similar future cases, separating functional fixes from technical bugs can help improve traceability, review, and deployment control.

Silvitui replied:
Thanks for the note! Just to clarify: the change from `*ngIf="solutionSent"` to `*ngIf="!solutionSent"` was related to an Angular warning (`NG0100: ExpressionChangedAfterItHasBeenCheckedError`) that appeared after submitting a solution.

It wasn’t a breaking error like the others, but it did show up in the console and was part of the overall goal of cleaning up the component behavior. Inverting the condition prevents the warning and improves stability by ensuring that the view doesn’t try to re-render based on outdated state.

Let me know if you'd prefer this to be separated into another PR, but I included it here since it was directly related to the console cleanup.

Silvitui replied:
About the changelog and version update ...I’ve already added an entry in CHANGELOG.md and updated the version to 3.1.44-RELEASE in package.json and package-lock.json, following the PATCH guidelines from the wiki. Let me know if I missed anything!
Thank you

ivilarop replied:
> Thanks for the note! Just to clarify: the change from `*ngIf="solutionSent"` to `*ngIf="!solutionSent"` was related to an Angular warning (`NG0100: ExpressionChangedAfterItHasBeenCheckedError`) that appeared after submitting a solution.
> 
> It wasn’t a breaking error like the others, but it did show up in the console and was part of the overall goal of cleaning up the component behavior. Inverting the condition prevents the warning and improves stability by ensuring that the view doesn’t try to re-render based on outdated state.
> 
> Let me know if you'd prefer this to be separated into another PR, but I included it here since it was directly related to the console cleanup.

If the behavior of the buttons does not change and it is an issue related to console errors, let's leave it in this PR, but it is important that it is detailed in the PR description so as not to lose traceability.

ivilarop replied:
I close the conversation myself so I can merge the PR, otherwise there won't be time to close it in this sprint.

Silvitui committed:
chore: update changelog and version to 3.1.44-RELEASE (Taiga #528)

Silvitui commented:
About the changelog and version update ...I’ve already added an entry in `CHANGELOG.md` and updated the version to `3.1.44-RELEASE` in `package.json` and `package-lock.json`, following the PATCH guidelines from the wiki. Let me know if I missed anything!
Thank you
